### PR TITLE
Migrated to kafka-client-go

### PR DESCRIPTION
**Pull request by ilian2233**
# Description

## What

This is part of the migration to MSK. Changes include:

- Update to the env vars with new groups and adding lag time for consumers;
- Combining the consumers for metadata and content topics;
- Updating kafka-client-go from v2 to v3;
- Updating health check and adding monitoring provided by kafka-client-go library;
- Removing the KafkaMessage struct;
- Adding separation of messages based on the `Message-Type` header.

## Why

[Jira ticket](https://financialtimes.atlassian.net/browse/UPPSF-3254)

## Scope and particulars of this PR (Please tick all that apply)

- [x] Tech hygiene (dependency updating & other tech debt)
- [ ] Bug fix
- [ ] Feature
- [ ] Documentation
- [ ] Breaking change
- [ ] Minor change (e.g. fixing a typo, adding config)

___
This Pull Request follows the rules described in our [Pull Requests Guide](https://github.com/Financial-Times/upp-docs/tree/master/guides/pr-guide)

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -62,35 +61,17 @@ func main() {
...
+		Desc:   "Kafka group id used for message consuming.",
+		EnvVar: "KAFKA_CONSUMER_GROUP",
 	})
-	kafkaProxyRoutingHeader := app.String(cli.StringOpt{
-		Name:   "kafkaProxyHeader",
-		Value:  "kafka",
-		Desc:   "Kafka proxy header - used for vulcan routing.",
-		EnvVar: "KAFKA_PROXY_HOST_HEADER",
+	consumerLagTolerance := app.Int(cli.IntOpt{
+		Name:   "consumer_lag_tolerance",
```
Comment (atanasdinov) [inline main.go:71]: It's a nitpick but all of the other variables are using camelCase for the `Name` field.

File: main.go [hunk 2]
```diff
@@ -146,6 +127,12 @@ func main() {
 		Desc:   "Space separated list with content types - to identify accepted content types.",
 		EnvVar: "WHITELISTED_CONTENT_TYPES",
 	})
+	kafkaAddress := app.String(cli.StringOpt{
+		Name:   "kafkaAddress",
+		Value:  "kafka:9092",
+		Desc:   "Address used by the queue producer to connect to Kafka",
```
Comment (atanasdinov) [inline main.go:133]: Let's rephrase this to indicate that this address is used by both the consumer and producer.

File: main.go [hunk 3]
```diff
@@ -198,8 +186,9 @@ func main() {
 			BrokersConnectionString: *kafkaAddress,
 			Topic:                   *combinedTopic,
 			Options:                 kafka.DefaultProducerOptions(),
+			ConnectionRetryInterval: time.Minute,
 		}
-		messageProducer := kafka.NewProducer(producerConfig, log, 0, time.Minute)
+		messageProducer := kafka.NewProducer(producerConfig, log)
```
Comment (atanasdinov) [inline main.go:191]: Seems like we've missed the closure of the two message producers so let's add those with this PR (similarly to how it's done for the consumer).

File: healthcheck_handler_test.go [hunk 4]
```diff
@@ -236,10 +236,18 @@ type mockConsumer struct {
...
 
-func (p *mockConsumer) ConnectivityCheck() (string, error) {
+func (p *mockConsumer) MonitorCheck() error {
 	if p.isConnectionHealthy {
-		return "", nil
+		return nil
 	}
 
-	return "", fmt.Errorf("error connecting to the queue")
+	return fmt.Errorf("error connecting to the queue")
```
Comment (atanasdinov) [inline healthcheck_handler_test.go:244]: Let's use a different error message and include a test case where the `MonitorCheck()` is the failing one.

File: healthcheck_handler.go [hunk 5]
```diff
@@ -133,7 +149,23 @@ func (h *HealthcheckHandler) checkIfInternalContentAPIIsReachable() (string, err
 	return ResponseOK, nil
 }
 
-func (h *HealthcheckHandler) checkIfKafkaIsReachable() (string, error) {
+func (h *HealthcheckHandler) checkIfKafkaIsReachableFromConsumer() (string, error) {
+	err := h.consumer.ConnectivityCheck()
+	if err != nil {
+		return "", err
+	}
+	return "", nil
```
Comment (atanasdinov) [inline healthcheck_handler.go:157]: It's better to include a non-nil status message. Both `ResponseOK` or a custom message would be fine.

File: healthcheck_handler.go [hunk 6]
```diff
@@ -49,18 +50,29 @@ func checkKafkaProducerConnectivity(h *HealthcheckHandler) health.Check {
...
+		Checker:          h.checkIfKafkaIsReachableFromConsumer,
 	}
 }
 
-func checkKafkaProxyConsumerConnectivity(h *HealthcheckHandler) health.Check {
+func monitorKafkaConsumers(h *HealthcheckHandler) health.Check {
 	return health.Check{
 		BusinessImpact:   "Can't process PostPublicationEvents and PostMetadataPublicationEvents messages. Indexing for search won't work.",
-		Name:             "Check connectivity to the kafka-proxy",
+		Name:             "Consumers are lagging",
```
Comment (atanasdinov) [inline healthcheck_handler.go:71]: That should be the name of the check itself so I'd suggest `Check consumer status`.

**Review by MiroslavGatsanoga — COMMENTED** [review]
Looks good, I am fine with approving the PR after the pending comments are addressed.

File: processor/msg_processor_test.go [hunk 7]
```diff
@@ -23,8 +23,49 @@ func testLogger() (*logger.UPPLogger, *hooks.Hook) {
...
+}
+
+func TestMsgProcessor_ProcessMessages_Closes_When_Channel_Closes(t *testing.T) {
+	ch := make(chan *kafka.FTMessage)
+	processor := MsgProcessor{src: ch}
+
+	timeout := time.After(3 * time.Second)
+	done := make(chan bool)
+
+	go func() {
```
Comment (atanasdinov) [inline processor/msg_processor_test.go:52]: Do we need this go routine wrapper?

File: healthcheck_handler.go [hunk 8]
```diff
@@ -49,18 +50,29 @@ func checkKafkaProducerConnectivity(h *HealthcheckHandler) health.Check {
...
+		Checker:          h.checkIfKafkaIsReachableFromConsumer,
+	}
+}
+
+func monitorKafkaConsumers(h *HealthcheckHandler) health.Check {
+	return health.Check{
+		BusinessImpact:   "Consumer is lagging behind when reading messages. Indexing of content will be delayed.",
+		Name:             "Check consumer status",
+		PanicGuide:       fmt.Sprintf("https://runbooks.ftops.tech/%s", systemCode),
+		Severity:         2,
```
Comment (atanasdinov) [inline healthcheck_handler.go:73]: The severity must be 3 (low).

**Review by atanasdinov — DISMISSED** [review]

**Review by atanasdinov — APPROVED** [review]

**Review by dtvalk-ov — APPROVED** [review]
